components/action-mgt/org.wso2.carbon.identity.action.management/src/test/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImplTest.java

```diff
@@ -0,0 +1,336 @@
+    public void testAddAction() throws ActionMgtException {
+
+        Action creatingAction = buildMockAction(
+                "PreIssueAccessToken",
+                "To configure PreIssueAccessToken",
+                "https://example.com",
+                buildMockBasicAuthentication("admin", "admin"));
+        action = serviceImpl.addAction(Action.ActionTypes.PRE_ISSUE_ACCESS_TOKEN.getPathParam(), creatingAction,
+                tenantDomain);
+        Assert.assertNotNull(action.getId());
+        Assert.assertEquals(creatingAction.getName(), action.getName());
+        Assert.assertEquals(creatingAction.getDescription(), action.getDescription());
+        Assert.assertEquals(Action.Status.ACTIVE, action.getStatus());
+        Assert.assertEquals(Action.ActionTypes.PRE_ISSUE_ACCESS_TOKEN.getActionType(),
+                action.getType().getActionType());
+        Assert.assertEquals(creatingAction.getEndpoint().getUri(), action.getEndpoint().getUri());
+        Assert.assertEquals(creatingAction.getEndpoint().getAuthentication().getType(),
+                action.getEndpoint().getAuthentication().getType());
+        Assert.assertEquals(creatingAction.getEndpoint().getAuthentication().getProperties().size(),
+                action.getEndpoint().getAuthentication().getProperties().size());
+    }
+
+    @Test(priority = 2, expectedExceptions = ActionMgtException.class,
+            expectedExceptionsMessageRegExp = "Unable to create an Action.")
+    public void testAddMaximumActionsPerType() throws ActionMgtException {
+
+        Action creatingAction = buildMockAction(
+                "PreIssueAccessToken",
+                "To configure PreIssueAccessToken",
+                "https://example.com",
+                buildMockBasicAuthentication("admin", "admin"));
+        action = serviceImpl.addAction(Action.ActionTypes.PRE_ISSUE_ACCESS_TOKEN.getPathParam(), creatingAction,
+                tenantDomain);
+    }
+
+    @Test(priority = 3)
+    public void testGetActionsByActionType() throws ActionMgtException {
+
+        List<Action> actions = serviceImpl.getActionsByActionType(
+                Action.ActionTypes.PRE_ISSUE_ACCESS_TOKEN.getPathParam(), tenantDomain);
+        Assert.assertEquals(1, actions.size());
+    }
+
+    @Test(priority = 4)
+    public void testGetActionsByActionTypeFromCache() throws ActionMgtException {
+
+        List<Action> actions = serviceImpl.getActionsByActionType(
+                Action.ActionTypes.PRE_ISSUE_ACCESS_TOKEN.getPathParam(), tenantDomain);
+        Assert.assertEquals(1, actions.size());
+    }
+
+    @Test(priority = 5)
+    public void testUpdateAction() throws ActionMgtException {
+
+        Action updatingAction = buildMockAction(
+                "Pre Issue Access Token",
+                "To update configuration pre issue access token",
+                "https://sample.com",
+                buildMockBasicAuthentication("updatingadmin", "updatingadmin"));
```

Comment:
Shall we change the authentication type also. Then we can assert whether the authentication properties are updated or not.

Reply:
But here the intention in this test case was to update characteristics of an action in the same action type and same authentication type. In a later test case updating of the authentication type happens.

Reply:
Updated in